sway-lib-std/src/experimental/storage.sw

```diff
@@ -0,0 +1,257 @@
+library;
+
+use ::alloc::{alloc, realloc_bytes};
+use ::hash::sha256;
+use ::option::Option;
+use core::experimental::storage::StorageHandle;
+
+/// Store a stack value in storage. Will not work for heap values.
+///
+/// ### Arguments
+///
+/// * `key` - The storage slot at which the variable will be stored.
+/// * `value` - The value to be stored.
+/// * `offset` - An offset, in words, from the beginning of slot at `key`, at which `value` should 
+///              be stored.
+///
+/// ### Examples
+///
+/// ```sway
+/// let five = 5_u64;
+/// write(ZERO_B256, 2, five);
+/// let stored_five = read::<u64>(ZERO_B256, 2).unwrap();
+/// assert(five == stored_five);
+/// ```
+#[storage(read, write)]
+pub fn write<T>(key: b256, offset: u64, value: T) {
+    // Get the number of storage slots needed based on the size of `T`
+    let number_of_slots = (offset * 8 + __size_of::<T>() + 31) >> 5;
+
+    // Allocate enough memory on the heap for `value` as well as any potential padding required due 
+    // to `offset`.
+    let padded_value = alloc::<u64>(number_of_slots * 32);
+
+    // Read the values that currently exist in the affected storage slots.
+    // NOTE: we can do better here by only reading from the slots that we know could be affected. 
+    // These are the two slots where the start and end of `T` fall in considering `offset`. 
+    // However, doing so requires that we perform addition on `b256` to compute the corresponding 
+    // keys, and that is not possible today.
+    let _ = __state_load_quad(key, padded_value, number_of_slots);
+
+    // Copy the value to be stored to `padded_value + offset`.
+    padded_value.add::<u64>(offset).write::<T>(value);
+
+    // Now store back the data at `padded_value` which now contains the old data but partially 
+    // overwritten by the new data in the desired locations.
+    let _ = __state_store_quad(key, padded_value, number_of_slots);
+}
+
+/// Reads a value of type `T` starting at the location specified by `key` and `offset`. If the 
+/// value crosses the boundary of a storage slot, reading continues at the following slot.
+///
+/// Returns `Option(value)` if a the storage slots read were valid and contain `value`. 
+/// Otherwise, return `None`.
+///
+/// ### Arguments
+///
+/// * `key` - The storage slot to load the value from.
+/// * `offset` - An offset, in words, from the start of slot at `key`, from which the value should 
+///              be read.
+///
+/// ### Examples
+///
+/// ```sway
+/// let five = 5_u64;
+/// write(ZERO_B256, 2, five);
+/// let stored_five = read::<u64>(ZERO_B256, 2);
+/// assert(five == stored_five);
+/// ```
+#[storage(read)]
+pub fn read<T>(key: b256, offset: u64) -> Option<T> {
+    // NOTE: we are leaking this value on the heap.
+    // Get the number of storage slots needed based on the size of `T`
+    let number_of_slots = (offset * 8 + __size_of::<T>() + 31) >> 5;
```

Comment:
Why `offset * 8` here?
Isn't `offset` already in words?

Reply:
`offset` is in words - multiplying it by 8 converts it to bytes which is the same unit as `__size_of::<T>()`.